### PR TITLE
Save resources obtained from ".get" in the snapshot

### PR DESCRIPTION
**Pull request by swgillespie**
Fixes https://github.com/pulumi/pulumi/issues/1521, prerequisite for https://github.com/pulumi/pulumi/issues/106.

Pulumi has had the static `.get` method exposed on `tfgen`-generated resources for some time now. This method ultimately forwards to the resource's constructor, but with the `id` field on `ResourceOptions` set. When `id` is set on `ResourceOptions`, the JavaScript language host invokes the engine using `ReadResource` to "read" the existing state of the resource and project it into the Pulumi object model by producing a fully-hydrated `Resource` object representing the existing object.

Because of #1521, it was not possible to directly or indirectly depend of a resource obtained from `ReadResource` because we did not insert resources that were read into the snapshot. As a result, `.get` was not useful.

This PR fixes the issue by aligning `ReadResource` a little bit more with `RegisterResource` in both semantics and implementation:

1. A new step is added to the engine: `Read`. `Read` takes a series of properties and an ID and produces a fully-hydrated Resource representing a real resource that exists with the given properties and ID. `Read` is unique among steps in that it runs during previews as long as the `ID` the `Read` step is given is known.
2. Two new step ops are added to the engine and display layer: `OpRead` and `OpReadReplace`. The precise semantics of these two step ops are documented below.
3. A new field is added to the protobuf `ReadRequest` sent to the engine: `dependencies`, which is a list of URNs representing incoming dependencies for the resource read.
4. A new bit is added to each resource's State: `External`, which indicates whether or not this resource is "external" to Pulumi and whether or not we own that resource's life cycle. A resource that is the target of a `OpRead` step is `External`, since it is not legal for us to delete it.

It is not legal for `Create` or `Update` steps to be issued against an `External` resource. It is legal for a `Delete` step to be issued against an `External` resource - the semantics are that the deletion of an `External` resource is simply removing it from the snapshot.

### Precise Semantics for `OpRead` and `OpReadReplace`

A step op of `OpRead` is issued by the engine in these cases:

1. A `ReadResource` occurs and the engine does not observe an old resource in the snapshot with the given URN. (Conceptually, a `read-create`). The engine inserts the new read resource state into the snapshot.
2. A `ReadResource` occurs and the engine does observe an old resource in the snapshot with the given URN, but that resource is marked `External`. (Conceptually, a `read-update`). The engine removes the existing resource state from the snapshot and inserts the new read state.

A step op of `OpReadReplace` is issued by the engine in these cases:

1. A `ReadResource` occurs and the engine does observe an old resource in the snapshot with the given URN, but that engine is *not* marked `External`. The engine pends deletion of the existing resource and inserts the new read state into the snapshot.

It is possible for a `External` resource to be the target of both `OpDelete` and `OpDeleteReplace` steps. The semantics of both operations are that the `External` resource is removed from the snapshot - the provider is not invoked at all.

### UI Considerations

The addition of a new step implies that `Read` now shows up as part of the CLI experience of a pulumi update. The current state of this PR is that:

* `Read`s always appear in the output, even if the read produces something that is identical to what is already in the snapshot.
* `ReadReplacement`s are rendered as a replacement.

Here's a short demo of me running through one of the tests I added, illustrating the CLI experience here:

https://asciinema.org/a/nj3rBlAz24CImj0vY5TEo4yUJ

Note that a `read-replacement` appears as part of the `details` view of the last update - this may be a bug depending on what we want to do here. In general, though, I think I'd prefer to handle specific UI tweaks after this PR has landed to avoid having this PR get super huge. I'm happy to make any improvements desired before merging, though.

### Individual Commits

This PR is split into four commits for ease of review:

1. https://github.com/pulumi/pulumi/commit/f5dde44632859846e1c562ff1db95552347bfa49 - This commit contains the changes to our protobuf definitions to accommodate the new field (`dependencies`) added to `ReadRequest`. The changes in this commit are auto-generated by `protoc` except for the changes to `resource.proto`.
2. https://github.com/pulumi/pulumi/commit/fc88ff6ba78db492a75505d14bc707a1cc336ed2 - This commit makes the CLI send V2 versions of `Resource` to the service. Since the service does not currently accept V2 `Resource`s, we should remove this commit before merging this PR. This was useful for development.
3. https://github.com/pulumi/pulumi/commit/ab5fdd1f34ca05a0fa8e35be23f1db16d04f123b - The meat of this PR. This commit contains all of the engine code required to surface `Read` as a new step.
4. https://github.com/pulumi/pulumi/commit/967a605cfbbb44123bef862f2c6b71a12c5862bd - This commit adds a few tests that test the corner cases of this PR - namely, the replacement of `External` resources with non-`External` resources and the other way around.

### CODE REVIEW

**Comment by swgillespie** [comment]
The last two commits fix the merge conflicts and address feedback that @pgavlin the other day about a "relinquish" operation - where the engine doesn't need to delete a resource that is being replaced by a read if the IDs of the old and new resources are identical. I've added a test for this new behavior as well.

Since this PR is the first to adopt the V2 breaking changes, this commit (https://github.com/pulumi/pulumi/pull/1654/commits/f60db6365b3310841566a91f64f9604bcefd4967) moves the CLI to use V2 resources throughout.

**Comment by swgillespie** [comment]
Re-triggered the Travis leg now that the service can handle V2 deployments.

**Comment by swgillespie** [comment]
With the last rebase and service changes today this PR should now be passing all new tests. I'd like to get this in over the next few days if possible.

**Comment by swgillespie** [comment]
@pgavlin @hausdorff I'd like to get this merged today if possible to unblock forward progress on https://github.com/pulumi/pulumi/issues/1626 and https://github.com/pulumi/pulumi/pull/1673, I'm sure you're both busy doing stuff for M16 but if you could spare a little bit to look over this I'd appreciate it.

**Comment by joeduffy** [comment]
Ok, I reread this a few times, and now think I am following. Just a few follow-ups to make sure.

> A step op of OpRead is issued by the engine in these cases:

Let me put some examples in here to check my understanding.

> A ReadResource occurs and the engine does not observe an old resource in the snapshot with the given URN. (Conceptually, a read-create). The engine inserts the new read resource state into the snapshot.

This happens the first time I run a Pulumi program containing a line like

```typescript
let vm = aws.ec2.Instance.read("my-vm", "i-b3476fe98");
```

> A ReadResource occurs and the engine does observe an old resource in the snapshot with the given URN, but that resource is marked External. (Conceptually, a read-update). The engine removes the existing resource state from the snapshot and inserts the new read state.

This happens all subsequent times I run that same Pulumi program.

> A step op of OpReadReplace is issued by the engine in these cases:

Here is where things get a little tricky.

> A ReadResource occurs and the engine does observe an old resource in the snapshot with the given URN, but that engine is _not_ marked External. The engine pends deletion of the existing resource and inserts the new read state into the snapshot.

This would happen if ran a Pulumi program containing this

```typescript
let vm = new aws.ec2.Instance("my-vm", { ... });
```

followed by editing that same Pulumi program and rerunning it as follows

```typescript
let vm = aws.ec2.Instance.read("my-vm", "i-b3476fe98");
```

The one case that I am unsure about is what if I do the reverse. In other words, start with

```typescript
let vm = aws.ec2.Instance.read("my-vm", "i-b3476fe98");
```

and then replace it with

```typescript
let vm = new aws.ec2.Instance("my-vm", { ... });
```

Does that simply perform an ordinary replacement operation and, because the old state is marked External, everything works out just right?

Thanks -- just want to make sure I get the flow here before diving in!

File: pkg/apitype/core.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ import (
 const (
 	// DeploymentSchemaVersionCurrent is the current version of the `Deployment` schema.
 	// Any deployments newer than this version will be rejected.
-	DeploymentSchemaVersionCurrent = 1
+	DeploymentSchemaVersionCurrent = 2
```
Comment (joeduffy) [inline pkg/apitype/core.go]: Does this necessitate a change to the service? (Or did it already happen?)

File: pkg/resource/deploy/source.go [hunk 2]
```diff
@@ -77,3 +77,22 @@ type RegisterResourceOutputsEvent interface {
 	// Done indicates that we are done with this step.  It must be called to perform cleanup associated with the step.
 	Done()
 }
+
+// ReadResourceEvent is an event that asks the engine to read the state of a resource that already exists.
+type ReadResourceEvent interface {
+	SourceEvent
+
+	ID() resource.ID
```
Comment (joeduffy) [inline pkg/resource/deploy/source.go]: Nit, missing comments on exported members.

File: pkg/resource/deploy/step.go [hunk 3]
```diff
@@ -102,6 +102,7 @@ func NewCreateStep(plan *Plan, reg RegisterResourceEvent, new *resource.State) S
 	contract.Assert(new.URN != "")
 	contract.Assert(new.ID == "")
 	contract.Assert(!new.Delete)
+	contract.Assert(!new.External)
```
Comment (joeduffy) [inline pkg/resource/deploy/step.go]: This just made me think of an alternative design. I'm not advocating it per se, however it could be simpler.

First, recall that we are likely to want to do a refresh operation before performing updates. See https://github.com/pulumi/pulumi-terraform/issues/57. As soon as we do that, we are going to be doing more "read"-like operations. It's worth considering how we will model those in the engine, given these changes.

Second, it occurs to me that we could actually just say that, instead of `OpRead` and `OpReadReplacement`, the existing operations `OpCreate`, `OpUpdate`, and `OpCreateReplacement` handle these functions. They would simply mean something different when `External` is `true`, just as `OpDelete` is different when `External` is `true`. The nice thing is that this may help inch us towards https://github.com/pulumi/pulumi-terraform/issues/57.

This is appealing to me because on the UI, it's not clear to me we want to expose these operations so prominently. So we'd just keep our existing visualizations and merely present them differently for external resources in a consistent way. I'm also sure *some* people will want to see them, but because they are external to the stack, it's probably far less interesting.  And, indeed, it may be alarming, since most people assume resources fetched in this manner are *not* under the management of Pulumi. If they see one getting deleted, they are likely to freak out. So I suspect we'd want to filter them out by default or summarize them in a specific way.

Did you consider this approach? I'm curious what you think about the tradeoffs involved?

File: pkg/resource/deploy/step_generator.go [hunk 4]
```diff
@@ -34,13 +34,65 @@ type stepGenerator struct {
...
+	old, hasOld := sg.plan.Olds()[urn]
+
+	// If the snapshot has an old resource for this URN and it's not external, we're going
+	// to have to delete the old resource and conceptually replace it with the resource we
+	// are about to read.
+	//
+	// We accomplish this through the "read-replacement" step, which atomically reads a resource
+	// and marks the resource it is replacing as pending deletion.
+	//
+	// In the event that the new "read" resource's ID matches the existing resource,
```
Comment (joeduffy) [inline pkg/resource/deploy/step_generator.go]: This surprised me. So, are you saying we won't physically delete the resource, when you change your program from `new`ing a resource to `get`ting it, provided the IDs match identically? This makes me a little nervous, as it's not just composition of existing operations.

File: pkg/resource/deploy/step_generator.go [hunk 5]
```diff
@@ -167,7 +223,28 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, err
 		}, nil
 	}
 
-	// Case 2: hasOld
+	// Case 2: wasExternal
```
Comment (joeduffy) [inline pkg/resource/deploy/step_generator.go]: Hmm, this is getting subtle.

I get that relinquish and taking ownership are super interesting and powerful conceptual operations, I just get nervous that automatically performing them when you change your program from `new` to `get` (or vice versa) is very subtle. I would have almost preferred explicit options to perform these actions and to by default do what the natural composition of operations would be.

Also, unless I'm mistaken, it really isn't taking ownership of an existing resource, because we're creating a replacement?

File: pkg/backend/cloud/client/client.go [hunk 6]
```diff
@@ -506,7 +506,7 @@ func (pc *Client) PatchUpdateCheckpoint(ctx context.Context, update UpdateIdenti
 	}
 
 	req := apitype.PatchUpdateCheckpointRequest{
-		Version:    1,
+		Version:    2,
```
Comment (joeduffy) [inline pkg/backend/cloud/client/client.go]: I'll admit, it wasn't evident from reading this change why we need to rev the formats. It would seem all changes to the deployment/checkpoint structures are purely additive.

File: pkg/resource/deploy/step_generator.go (hunk 4 above)
Comment (lukehoban) [reply pkg/resource/deploy/step_generator.go]: Yes - this is a new case, and something @pgavlin @swgillespie and I spoke about in person at length this week.  

There are two things that convince me this design makes sense:
1. It supports a really valuable scenario - detach management of an existing resource from Pulumi without destroying the resource.  For example, this would be useful when splitting one Pulumi stack into multiple stacks and factoring some resources out of the original stack to be `.gets`.
2. If we didn't do this special case behaviour, the operation would be guaranteed to fail - we would delete the resource immediately before (or worse, immediately after - or even worse, non-deterministically before/after) the `.get` tries to capture a handle to it.  It's hard to imagine what benefit it provides to allow this to happen, especially given that (1) is a more useful interpretation.

File: pkg/resource/deploy/step_generator.go (hunk 5 above)
Comment (lukehoban) [reply pkg/resource/deploy/step_generator.go]: > I get that relinquish and taking ownership are super interesting and powerful conceptual operations, I just get nervous that automatically performing them when you change your program from new to get (or vice versa) is very subtle. I would have almost preferred explicit options to perform these actions and to by default do what the natural composition of operations would be.

I understand the nervousness, but I'll admit that as soon as we introduce this notion of first class reads, we kind of have to provide semantics for these cases, and I think the semantics here are the most expected ones if you think about what a user might expect to happen when making these changes.

Having these behave in a surprising way and offering another tool to do what is expected doesn't necessarily feel better, even if it could possibly simplify the model here.

I think the complexity here is somewhat fundamental to having a 1st class notion of "external" resources in a resource model.

File: pkg/backend/cloud/client/client.go (hunk 6 above)
Comment (lukehoban) [reply pkg/backend/cloud/client/client.go]: Purely additive changes still require revving the checkpoint format if the interpretation of a resource by an old engine would be incorrect.  In this case, an old engine would ignore "external" and would then make incorrect decisions about how to interpret it.  Had we stored these "external" resources in a separate data structure in the checkpoint file, we may have been able to do this in a way which didn't have this problem, but with the approach we did take, it was necessary to rev the format.

File: pkg/resource/deploy/step.go [hunk 7]
```diff
@@ -391,6 +397,103 @@ func (s *ReplaceStep) Apply(preview bool) (resource.Status, error) {
...
+func (s *ReadStep) Res() *resource.State { return s.new }
+func (s *ReadStep) Logical() bool        { return !s.replacing }
+
+func (s *ReadStep) Apply(preview bool) (resource.Status, error) {
+	urn := s.new.URN
+	id := s.new.ID
+
+	// Unlike most steps, Read steps run during previews. The only time
+	// we can't run is if the ID we are given is unknown.
+	if id != plugin.UnknownStringValue {
```
Comment (pgavlin) [inline pkg/resource/deploy/step.go]: This might need to be `if id != ""`. At least in the case of `RegisterResource`, the `id` isn't a typical `PropertyValue`, it's just a string.

File: pkg/resource/stack/deployment.go [hunk 8]
```diff
@@ -90,19 +93,65 @@ func DeserializeDeployment(deployment *apitype.UntypedDeployment) (*deploy.Snaps
...
+		}
+	default:
+		contract.Failf("unrecognized version: %d", deployment.Version)
+	}
+
+	return DeserializeDeployment(v2deployment)
+}
+
+// DeserializeDeployment deserializes a typed Deployment into a `deploy.Snapshot`.
+func DeserializeDeployment(deployment apitype.DeploymentV2) (*deploy.Snapshot, error) {
```
Comment (pgavlin) [inline pkg/resource/stack/deployment.go]: Nit: `DeserializeDeploymentV2`

File: pkg/resource/deploy/step_generator.go (hunk 5 above)
Comment (lukehoban) [reply pkg/resource/deploy/step_generator.go]: Erroring in this case unless you pass `convertExistingResource: true` seems reasonable.  Though not clear that this helps with the issue of deleting the resource if you got the ID wrong.  And overall, feels like it puts a speed bump on the wrong side of the usability here - the bad thing is just as easy, the good thing is just harder.

I'm sympathetic to composition arguments, but once we introduce new first class notion of external resources, the state transition matrix just isn't necessarily orthogonally compositional.  

We should probably have offline discussion on this to make sure we are all on the same page - I know @swgillespie and @pgavlin spent several hours working through this exact topic - and then another hour with me earlier this week.

File: pkg/apitype/core.go (hunk 1 above)
Comment (swgillespie) [reply pkg/apitype/core.go]: This already happened earlier this week.

File: pkg/resource/deploy/step_generator.go (hunk 5 above)
Comment (pgavlin) [reply pkg/resource/deploy/step_generator.go]: One of the other issues with an approach like `convertExistingResource: true` is that once that program has executed, it may immediately be incorrect, assuming that `convertExistingResource` expects the current state to be non-external. This would then require the user to make another code change in order to make forward progress.

File: pkg/backend/cloud/client/client.go (hunk 6 above)
Comment (pgavlin) [reply pkg/backend/cloud/client/client.go]: To restate: a change that is purely additive in a structural sense requires a version bump if its semantics are not backwards-compatible. First-class providers are another instance of this, which is why we folded the schema changes to that feature into final version of `ResourceV2` et. al.

File: pkg/resource/deploy/step.go (hunk 3 above)
Comment (pgavlin) [reply pkg/resource/deploy/step.go]: @swgillespie and I discussed this approach a bit last week. Personally, I am in favor of something like what you describe: my gut says that the resulting code will be a bit simpler. We agreed to go ahead with the current strategy, though, and make a final decision after seeing how things turned out.

Comment (swgillespie) [reply pkg/resource/deploy/step.go]: Pat had the same (or a similar) idea when we whiteboarded this last week. Here are my thoughts:

It is possible to implement refresh in terms of both `OpRead` and `OpCreate` and friends. In both cases, however, it is difficult to implement https://github.com/pulumi/pulumi-terraform/issues/57, because the step generator must know the results of the refresh before deciding what kind of step to generate for the resource itself. This is in contrast to the step generator's normal mode of operation, where steps are generated using only the knowledge of incoming new inputs and existing old inputs. Therefore, this implementation strategy is less of a question about implementation efficiency and correctness but more of a question of software engineering, to steal Pat's quote.

I lean towards having small, composable steps whose semantics are small over larger steps whose semantics are more complex. The reason for this is that it pushes complexity away from step application and towards the step generator, where we can centralize this logic in one place that can be tested as a single unit. If the outputs of the step generator (steps) are small and have well-known and simple semantics, in my mind it is easier to reason about the correctness of a plan as a whole because it is clear just by looking at the steps themselves (and not their inputs) what a particular step will do.

Having `OpCreate`, `OpUpdate`, `OpCreateReplacement` change semantics depending on whether or not their target(s) are `External` flies a little bit in the face of that tendency I have. When looking at a series of steps generated by the step generator, one would also have to know the target resources (whether or not they were `External`) to fully understand the semantics of the step.

That said, the fact that `OpDelete` handles both `External` and non-`External` resources in this PR is a contradiction to everything I have just said. I actually did have a `read-delete` step in a previous version of this PR. Here's how I justify it, though: the primitives of Pulumi are read, update, create, and delete. All operations can be phrased in terms of those four operations. In my mind it makes sense to conflate `OpDelete` with the deletion of an external resource, because conceptually we're still *deleting* - we're just removing something from a snapshot. However, in my mind it does *not* make sense to conflate `read` and `create`, since fundamentally those are two very different operations that bottom out at different provider operations.

File: pkg/resource/deploy/step_generator.go (hunk 5 above)
Comment (swgillespie) [reply pkg/resource/deploy/step_generator.go]: Early on when doing this PR I thought that we might want to consider some sort of CLI gesture when doing one of these more advanced operations. I agree with Pat that this becomes troublesome when the gesture enters the code because we're drifting away from desired state configuration semantics. Luke introduced a test that's been very helpful when we were talking last week - any two sequences of steps through our system composed in any order should get us to the same desired state ("path independence").

**Comment by swgillespie** [comment]
thanks everyone!
